### ISSUE TITLE
Stripe volume from saved filename as an option

### ISSUE THREAD

**Issue by GlassedSilver** [fmd-project-team/FMD#893]
Hello! :)

First of all I wanna start out by saying that I'm a massive fan of this software and it makes life so much better. I cannot be thankful enough, however I have a little bit of a predicament here:

Whilst I realize that there is a sorting number I can prefix the output filename with to make the saved chapters appear in the correct order in the file manager there is the issue that many manga are sometimes picked up by different scanlation groups throughout their run, resulting in inconsistent naming of these chapters. Some groups mention the volume, some don't.

Since I store my manga in the `Manga name / Chapter nn - Chapter title.cbz` pattern this results in some files starting with `Vol. 2 Ch. 12`, some `Ch. 9` etc... You know what I mean.

The sorting number is a nice workaround, unless of course you want a) cleaner filenames where you don't even care about the volume number, but prefer the improved readability of just the chapter number along with the chapter title and b) and this is more crucial: I want to store my manga in the new and emerging comic/manga server [Komga](https://github.com/gotson/komga/). Komga sorts chapters (internally "books" in a series) by filename and this is how it will serve the chapters to client software like Tachiyomi. (there is a Komga extension for Tachiyomi, yes. Using Tachiyomi's automatic updating of my MAL/Anilist and Kitsu reading progress I'm basically in reading heaven! :D )

However, if I include the sorting number as prefix Komga will use this number as chapter number. This all goes up in flames if there is a chapter 0 or "half-chapters" (16.5 for example) or alt versions of a chapter...

Having chapters only as the very first thing of the file name along with, if we can implement this, a consistent descriptor for chapters (like "always rename to Ch." in case a group writes "Chapter 9" instead of the more ubiquitous "Ch.") then I'd be in even higher heavens. We'd get a lot closer to having a solid software stack for manga reading. (the stack being: downloader, watcher included, server/client reader, tracking progress)

**Comment by dazedcat19** [comment]
fmd is basically just a website crawler. we have no control of the name. thought we can modify it to some extent on delivery. it will not cover all thr possible name variant for many website.
if you really want to, you can edit the lua file and change it as you please.

**Comment by GlassedSilver** [comment]
> fmd is basically just a website crawler.
Yes, right... I am well aware FMD doesn't run hundreds of scanlation sites and oversee what must be thousands of scanlation groups.

> we have no control of the name.
True for what is uploaded, not for the final filename which is all I care about obviously? Maybe me explaining my planned software stack scenario confused you? When I mention Komga as server that means a local-only server that I use to host my downloads in my own LAN.

> thought we can modify it to some extent on delivery. it will not cover all thr possible name variant for many website.
If the logic gets implemented in a setting that allows customization for search patterns then we could shoot at the 95% of most common phrases and allow for customization in case some oddballs arise for a few series. Right now I'm looking at all manual work...

> if you really want to, you can edit the lua file and change it as you please.
And have it overwritten by the application updating it itself? Regardless, why not have a setting for this?

**Comment by dazedcat19** [comment]
you can try to implement it and make a pull request. with some quality review it might be accepted.
if you make a copy of the lua file and rename it so it will be sorted last. something like `z_mangastream.lua` it will not be overwritten.

**Comment by GlassedSilver** [comment]
> you can try to implement it and make a pull request. with some quality review it might be accepted.
> if you make a copy of the lua file and rename it so it will be sorted last. something like `z_mangastream.lua` it will not be overwritten.

Not really that much of a programmer, but I can try... Never dabbled with Lua, in any case it'd be helpful if the UI part of it would be implemented by someone else, I will merrily supply the logic like hunting down frequently used wordings and long-term support of adding new ones. (I'd then be open to having assigned an issue for excludables)

Not too much of a fan of just putting a z_xyz.lua, because that probably implies that a) I have to manually do that to every lua and b) I would need to diff and merge updates myself every time there are official updates to the luas. They aren't updated all the time for nothing after all?

**Comment by dazedcat19** [comment]
No. Lua is the embedded script / plugins. You should implement it in fmd. maybe with ui. or use existing one. like the checkbox for replace volume. uncheck it will exclude the volume.

**Comment by GlassedSilver** [comment]
Maybe I'm just too thick to get it, maybe the code architecture is just something I'm not even remotely enough aware of, but I cannot for the life of me find a place to begin.

I wouldn't even know what to write into the lua file to define the output, especially as the file naming logic seems to be defined in 'config/config.ini'??

Honestly, I'd love to help with a PR, but I'm looking at a blank sheet of paper here. :(

Any assistance is much appreciated.

**Comment by GlassedSilver** [comment]
What I just noticed thinking about this a little more is that Tachiyomi for example might offer nice insight into how to code this.

Tachiyomi does allow downloading from Mangadex (to just speak from my main site experience) and even with wonky filenames it will parse the chapter names properly and update chapter 4 as good as a chapter 9.5. So it definitely reads those correctly and seems to ignore volumes.

Not sure if it helps, thought I'd add my experience with Tachiyomi though since it seems to do something pretty well that allows it to automatically recognize chapter numbers and not being thrown off by volumes being there or not.

If there is anything else I can add to this issue without proper coding experience in the given languages I'll be happy to help.

**Comment by dazedcat19** [comment]
It's in `uBaseunit.pas` method `CustomRename`. The parsing for volume/chapter is in `uMisc.pas` method `VolumeChapterPadZero`. FMD recognize no volume, when there's only one number sequence detected with no vol/volume prefix. But the chapter name it receive must be on some standard. That must be done in the specific modules. The parsing itself may need improvement indeed.
What you have to do:
- Check the UI and its stored value. If I remembered correctly, the rename vol option is the parent of rename chapter. So you have to break the parent-child relationship to something equal. 
- Add new method to parse the chapter name. Split the vol number, chapter number and title of chapter. `CustomRename` will use this method.
- Change the `CustomRename` method to respect both option indivually. If checkbox rename vol is true then include the volume number, the same with chapter number. if option is unchecked then omitt them.
- Remove the method `VolumeChapterPadZero`, and replace it with your own if necessary. 
- Check the modules you're targeting, format the chapter before feeding it to `CustomRename` if necessary.

**Comment by GlassedSilver** [comment]
Sorry, I have neither of those files?

I am completely lost, sorry to cause this much irritation, but I think either a) this isn't ELI5 enough for my skill level or b) I am lacking some files due to a botched update maybe?

**Comment by dazedcat19** [comment]
it's right in the `baseunits` folder. 
https://github.com/fmd-project-team/FMD/blob/master/baseunits/uBaseUnit.pas
https://github.com/fmd-project-team/FMD/blob/master/baseunits/uMisc.pas
